### ISSUE TITLE
Deferred cancellation documentation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4320]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#4320|
|Type|defect|
|Created|2010-02-26 10:12:04Z|
|Branch|https://github.com/twisted/twisted/tree/documenta-4320|

There is no mention of cancellation here: http://twistedmatrix.com/documents/current/core/howto/defer.html

<details><summary>Searchable metadata</summary>

```
trac-id__4320 4320
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__Twisted_13_1_0 Twisted-13.1.0
branch__branches_documenta_4320 branches/documenta-4320
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1267179124000000 1267179124000000
changetime__1376999061000000 1376999061000000
version__None None
owner__kaizhang kaizhang
cc__exarkun cc__thijs cc__kaizhang
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @jesstess|
|-|-|

I think we might want to consider this a priority for 10.1.  Perhaps not an absolute requirement, but in general we should be trying harder to introduce documentation closer to the time when we introduce new features.  And `Deferred` is a serious core-level chunk of functionality of Twisted, so increasing its functionality without increasing its documentation is disproportionately bad.

Most likely glyph has plenty of other things on his plate.  Optimistically assigning this to someone who appears to have more Twisted hacking time available these days.  I'm available to help out and answer any questions which might arise, too.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

Twisted 10.1.0 has been released. Removing this ticket from the 10.1 milestone. If you think it's severe enough to demand a follow-up release, please say so and add this ticket back to the 10.1 milestone.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

I know tubes are super exciting and everything, but this is a somewhat glaring oversight that is going to start getting embarassing soon.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_8):
> I know tubes are super exciting and everything, but this is a somewhat glaring oversight that is going to start getting embarassing soon.

I think there is at least one false dichotomy in this comment, but thanks for the reminder nonetheless.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [36229]) Branching to 'documenta-4320'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

It should still be in the next milestone.  Let's try to get it done this time...

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_13"></a>|@hynek commented|
|-|-|

Another milestone came and went. So there is

 - http://blog.futurefoundries.com/2012/11/cancelling-deferreds.html
 - http://blog.futurefoundries.com/2012/11/cancelling-deferreds-part-2.html
 - http://blog.futurefoundries.com/2013/01/deferred-cancellation-part-3-timeouts.html

I understand we also have a GSoC student working on better cancellation – shouldn’t it be manageable for a native speaker to convert those three blog posts into a doc page?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Itamar, what do you think of turning those blog posts into documentation?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst commented|
|-|-|

I'm happy to grant the right to Twisted to use it under MIT license.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Hynek, if you can do the formatting mucking about to get that stuff into lore documents, I will volunteer for native-speaker review :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Feel free to just make a complete mess of it, as long as it's valid lore; no segues or anything; I can fill in any gaps as I proofread.

Also, keep in mind there's already a bit of a document in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_18"></a>|@hynek commented|
|-|-|

Before there’s another release w/o, I would do it. But wouldn’t it make more sense if our GSoC student who currently probably knows cancellation best of us all did it? Including the stuff that has been done *after* the blog posts?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_18):
> Before there’s another release w/o, I would do it. But wouldn’t it make more sense if our GSoC student who currently probably knows cancellation best of us all did it? Including the stuff that has been done *after* the blog posts?

I think it would make sense for you to work *with* our GSoC student :).  Whose name, by the way, is Zhang Kai, or `kaizhang` here on Trac.

Hi Zhang.  Meet Hynek :).

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|kaizhang commented|
|-|-|

Replying to [glyph](#note_19):
> Replying to [hynek](#note_18):
> > Before there’s another release w/o, I would do it. But wouldn’t it make more sense if our GSoC student who currently probably knows cancellation best of us all did it? Including the stuff that has been done *after* the blog posts?
> 
> I think it would make sense for you to work *with* our GSoC student :).  Whose name, by the way, is Zhang Kai, or `kaizhang` here on Trac.
> 
> Hi Zhang.  Meet Hynek :).

Hi, I'm glad to help on this ticket. I will talk with my mentors(itamar and ashfall) about this.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|kaizhang commented|
|-|-|

I have converted itamar's blog posts into the Deferred cancellation documentation. I didn't add the documentation about timeout because we haven't finished it(ticket [[#5786](https://github.com/twisted/twisted/issues/5786)](https://github.com/twisted/twisted/issues/5786)). We can add it in another ticket after ticket #5786 is finished. The same for the DeferredList cancellation documentation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

I forced a build.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph set owner to kaizhang<br>@glyph set status to `new`|
|-|-|

I made a slight change to the text to address the build failure described on [#6683](https://github.com/twisted/twisted/issues/6683).

I think that this should land more or less as-is, since it adds some much-needed documentation.  There are ways that I could think of it being improved, but I can file those as enhancement bugs later - there's no reason not to land this that I can see :).

The documentation for [[#5786](https://github.com/twisted/twisted/issues/5786)](https://github.com/twisted/twisted/issues/5786) should be added _in_ #5786; this ticket being separate from the initial implementation of Deferred cancellation was a mistake, and generally we should add documentation in the same change that adds a feature.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|kaizhang set status to `closed`|
|-|-|

(In [39643]) Merge documenta-4320: Added Deferred cancellation documentation

Author: glyph, kaizhang
Reviewer: glyph
Fixes: [#4320](https://github.com/twisted/twisted/issues/4320)

Added Deferred cancellation documentation in howto page of Deferred.
